Sming/Wiring/IpAddress.h

```diff
@@ -47,75 +53,83 @@ class IpAddress : public Printable
 	{
 	}
 
-	IpAddress(uint8_t first_octet, uint8_t second_octet, uint8_t third_octet, uint8_t fourth_octet)
+	IpAddress(ip_addr_t &addr)
 	{
-		IP4_ADDR(&address, first_octet, second_octet, third_octet, fourth_octet);
+		address = addr;
 	}
 
-	IpAddress(uint32_t address)
+	IpAddress(const ip_addr_t &addr)
 	{
-		this->address.addr = address;
+		address = addr;
 	}
 
-	IpAddress(ip_addr address)
+#if LWIP_VERSION_MAJOR == 2 && LWIP_IPV6
+	IpAddress(ip4_addr_t &addr)
 	{
-		this->address.addr = address.addr;
+		ip_addr_copy_from_ip4(address, addr);
 	}
+#endif
 
-#if LWIP_VERSION_MAJOR == 2
-	IpAddress(ip_addr_t address)
+	IpAddress(uint8_t first_octet, uint8_t second_octet, uint8_t third_octet, uint8_t fourth_octet)
 	{
-		this->address = address;
+		IP_ADDR4(&address, first_octet, second_octet, third_octet, fourth_octet);
+	}
+
+	IpAddress(uint32_t address)
+	{
+		ip_addr_set_ip4_u32(&this->address, address);
 	}
-#endif
 
 	IpAddress(const uint8_t* address)
 	{
-		IP4_ADDR(&this->address, address[0], address[1], address[2], address[3]);
+		IP_ADDR4(&this->address, address[0], address[1], address[2], address[3]);
 	}
 
 	IpAddress(const String& address)
 	{
 		fromString(address);
 	}
 
-	// Overloaded cast operator to allow IpAddress objects to be used where a pointer
-	// to a four-byte uint8_t array is expected
 	operator uint32_t() const
 	{
-		return address.addr;
+		return ip_addr_get_ip4_u32(&address);
 	}
 
-	operator ip_addr() const
+	operator ip_addr_t() const
 	{
-		return {address.addr};
+		return address;
 	}
 
-	operator ip_addr*()
+	operator ip_addr_t*()
 	{
-		return reinterpret_cast<ip_addr*>(&address);
+		return &address;
 	}
 
-#if LWIP_VERSION_MAJOR == 2
-	operator ip_addr_t*()
+#if LWIP_VERSION_MAJOR == 2 && LWIP_IPV6
+	operator ip4_addr_t const()
 	{
-		return &address;
+		return *ip_2_ip4(&address);
+	}
+
+	operator ip4_addr_t*()
+	{
+		return ip_2_ip4(&address);
 	}
 #endif
 
-	bool operator==(const IpAddress& addr) const
+	bool operator==(const IpAddress& otherAddress) const
 	{
-		return ip_addr_cmp(&address, &addr.address);
+		return ip_addr_cmp((ip_addr_t *)&address, (ip_addr_t* )&otherAddress);
 	}
 
 	bool operator==(const uint8_t* addr) const
 	{
 		return *this == IpAddress(addr);
 	}
 
-	bool operator!=(const IpAddress& addr) const
+	bool operator!=(const IpAddress& otherAddress) const
 	{
-		return !ip_addr_cmp(&address, &addr.address);
+		return !ip_addr_cmp((ip_addr_t *)&address, (ip_addr_t* )&otherAddress);
```

Comment:
Could write as `return !ip_addr_cmp(&address, &otherAddress.address);` to avoid cast